### PR TITLE
Collection homework 1.0

### PR DESCRIPTION
**Pull request by vitalibutnar**
Add implementation for Student Set

### CODE REVIEW

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 1]
```diff
@@ -1,67 +1,133 @@
...
+    static final int DEFAULT_INITIAL_CAPACITY = 16;
+    static final float DEFAULT_LOAD_FACTOR = 0.75f;
+
+    private Node<E>[] set;
+    private int size;
+
+
+    @SuppressWarnings({"unchecked"})
+    StudentSet() {
+        set = (Node<E>[]) new Node[DEFAULT_INITIAL_CAPACITY];
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentSet.java:22]: Unnecessary cast to Node<E>[], can be removed.

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 2]
```diff
@@ -1,67 +1,133 @@
...
+        return new StudentSetIterator();
     }
 
     @Override
     public Object[] toArray() {
-        //TODO
-        return new Object[0];
+        int q = 0;
+        Object[] result = new Object[size];
+        Iterator<E> iter = iterator();
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentSet.java:51]: Can use a for loop instead:
for (final E e : this) {
            result[q++] = e;
        }

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 3]
```diff
@@ -1,67 +1,133 @@
...
+        }
+        return es;
     }
 
     @Override
-    public boolean add(Student student) {
-        //TODO
-        return false;
+    public boolean add(E t) {
+        int hash = hash(t), index = (set.length - 1) & hash;
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentSet.java:70]: Declare variable on a separate line
https://www.oracle.com/java/technologies/javase/codeconventions-declarations.html

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 4]
```diff
@@ -1,67 +1,133 @@
...
+        } while (node != null);
         return false;
     }
 
+    @SuppressWarnings({"unchecked"})
     @Override
     public void clear() {
-        //TODO
+        if (size > 0 && set != null) {
+            set = (Node<E>[]) new Node[set.length];
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentSet.java:122]: Unnecessary casting Node<E>[]

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 5]
```diff
@@ -1,67 +1,133 @@
...
+            size = 0;
+        }
     }
 
     @Override
-    public boolean addAll(Collection<? extends Student> collection) {
-        //TODO
-        return false;
+    public boolean addAll(Collection<? extends E> collection) {
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/collections/StudentSet.java:128]: Could be shorter: collection.stream().anyMatch(this::add);

Comment (vitalibutnar) [reply src/main/java/com/endava/internship/collections/StudentSet.java:128]: Then it won't add all the elements. anyMatch work before first match

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 4 above)
Comment (vitalibutnar) [reply src/main/java/com/endava/internship/collections/StudentSet.java:122]: Ok, I've got it. I'll fix it in next Push

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 3 above)
Comment (vitalibutnar) [reply src/main/java/com/endava/internship/collections/StudentSet.java:70]: I'll fix it in next Push

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 5 above)
Comment (vermurachi) [reply src/main/java/com/endava/internship/collections/StudentSet.java:128]: Then it won't work as it is now either, should try allMatch

Comment (vitalibutnar) [reply src/main/java/com/endava/internship/collections/StudentSet.java:128]: Ok, I'll fix it by forEach

**Pull request merged**